lib/oidcstrategy.js

```diff
@@ -836,60 +955,47 @@ Strategy.prototype._authCodeFlowHandler = function authCodeFlowHandler(code, tup
   let libraryProduct = aadutils.getLibraryProduct();
   let libraryProductParameterName = aadutils.getLibraryProductParameterName();
 
-  const config = self._options._configForOauth;
-
   const oauth2 = new OAuth2(
-    config.clientID, // consumerKey
-    config.clientSecret, // consumer secret
+    oauthConfig.clientID, // consumerKey
+    oauthConfig.clientSecret, // consumer secret
     '', // baseURL (empty string because we use absolute urls for authorize and token paths)
-    config.authorizationURL, // authorizePath
-    config.tokenURL, // accessTokenPath
+    oauthConfig.auth_endpoint, // authorizePath
+    oauthConfig.token_endpoint, // accessTokenPath
     {libraryProductParameterName : libraryProduct,
      libraryVersionParameterName : libraryVersion} // customHeaders
   );
 
-  let callbackURL = config.callbackURL;
-  // options.callbackURL is merged into config object while `setOptions` call
-  if (!callbackURL) {
-    return next(new Error('no callbackURL found'));
-  }
-
-  const parsedCallbackURL = url.parse(callbackURL);
-  if (!parsedCallbackURL.protocol) {
-    // The callback URL is relative, resolve a fully qualified URL from the
-    // URL of the originating request.
-    callbackURL = url.resolve(aadutils.originalURL(req), callbackURL);
-  }
-
   return oauth2.getOAuthAccessToken(code, {
     grant_type: 'authorization_code',
-    redirect_uri: callbackURL,
-  }, (getOAuthAccessTokenError, access_token, refresh_token, params) => {
-    if (getOAuthAccessTokenError) {
-    return next(new InternalOAuthError('failed to obtain access token', getOAuthAccessTokenError));
-    }
-
-    var id_token = params.id_token;
+    redirect_uri: oauthConfig.callbackURL,
+    scope: oauthConfig.scope
+  }, (getOAuthAccessTokenError, access_token, refresh_token, items) => {
+    if (getOAuthAccessTokenError)
+      return self.failWithLog(`In _authCodeFlowHandler: failed to redeem authorization code: ${aadutils.getErrorMessage(getOAuthAccessTokenError)}`);
 
     // id_token should be present
-    if (!id_token)
-      return self.fail('id_token is not received');
+    if (!items.id_token)
+      return self.failWithLog('In _authCodeFlowHandler: id_token is not received');
 
     // token_type must be 'Bearer'
-    if (params.token_type !== 'Bearer') {
-      log.info('token_type received is: ', params.token_type);
-      return self.fail(`token_type received is not 'Bearer'`);
+    if (items.token_type !== 'Bearer') {
+      log.info('token_type received is: ', items.token_type);
+      return self.failWithLog(`In _authCodeFlowHandler: token_type received is not 'Bearer'`);
     }
 
-    log.info('received id_token: ', id_token);
+    log.info('received id_token: ', items.id_token);
```

Comment:
why not? logging is only seen by the developers, with the id_token it is much easier for them to debug issues